src/java.base/share/classes/sun/security/pkcs12/PKCS12KeyStore.java

```diff
@@ -1479,25 +1481,32 @@ private byte[] calculateMac(char[] passwd, byte[] data)
         final PBEParameterSpec params;
         final String algName;
         final MacData macData;
-        final String kdfHmac;
+        String kdfHmac;
         int writeIterationCount = macIterationCount;
 
-        if (macAlgorithm.startsWith("PBEWith") ||
-                defaultMacAlgorithm().startsWith("PBEWith")) {
-            if (defaultMacAlgorithm().equals("PBEWithHmacSHA512")) {
-                kdfHmac = "HmacSHA512";
-            } else if (defaultMacAlgorithm().equals("PBEWithHmacSHA256")) {
-                kdfHmac = "HmacSHA256";
+        if (newKeystore) {
+            if (macAlgorithm.startsWith("PBEWith") ||
+                    defaultMacAlgorithm().startsWith("PBEWith")) {
+                kdfHmac = defaultMacAlgorithm().replace("PBEWith", "");
+                if (!(kdfHmac.equals("HmacSHA512") ||
+                        kdfHmac.equals("HmacSHA256"))) {
+                    kdfHmac = pbmac1Hmac; // use value associated with keystore
```

Comment:
`kdfHmac` is probably null now. If you decide to reject other algorithms (which I don't find necessary), error out.

Reply:
Why is `pbmac1Hmac` necessary? Do we want to support writing with different PRF and HMAC?

Reply:
Writing with different PRF and HMAC **is** supported. Test vector #2 in the RFC passes.
`pbmac1Hmac` is gone.